### PR TITLE
mimic: rgw: swift object expiry fails when a bucket reshards

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/39741

---

backport of #27870
parent tracker https://tracker.ceph.com/issues/39495

### CODE REVIEW

**Comment by smithfarm** [comment]
@theanalyst Whelp

File: src/rgw/rgw_bucket.cc [hunk 1]
```diff
@@ -502,7 +502,7 @@ static bool bucket_object_check_filter(const string& oid)
   return rgw_obj_key::oid_to_key_in_ns(oid, &key, ns);
 }
 
-int rgw_remove_object(RGWRados *store, RGWBucketInfo& bucket_info, rgw_bucket& bucket, rgw_obj_key& key)
+int rgw_remove_object(RGWRados *store, const RGWBucketInfo& bucket_info, const rgw_bucket& bucket, rgw_obj_key& key)
```
Comment (smithfarm) [inline src/rgw/rgw_bucket.cc]: @theanalyst This line appears to cause build failure:

```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_bucket.cc: In function 'int rgw_remove_object(RGWRados*, const RGWBucketInfo&, const rgw_bucket&, rgw_obj_key&)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_bucket.cc:515:82: error: passing 'const RGWBucketInfo' as 'this' argument discards qualifiers [-fpermissive]
   return store->delete_obj(rctx, bucket_info, obj, bucket_info.versioning_status());
                                                                                  ^
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_rados.h:17,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_bucket.cc:18:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_common.h:1370:7: note:   in call to 'int RGWBucketInfo::versioning_status()'
   int versioning_status() { return flags & (BUCKET_VERSIONED | BUCKET_VERSIONS_SUSPENDED | BUCKET_MFA_ENABLED); }
       ^~~~~~~~~~~~~~~~~
src/rgw/CMakeFiles/rgw_a.dir/build.make:234: recipe for target 'src/rgw/CMakeFiles/rgw_a.dir/rgw_bucket.cc.o' failed
make[3]: *** [src/rgw/CMakeFiles/rgw_a.dir/rgw_bucket.cc.o] Error 1
```

**Comment by theanalyst** [comment]
looks correct

**Comment by yuriw** [comment]
@smithfarm needs rebase

--- pr 29258 --- pulling https://github.com/smithfarm/ceph.git branch wip-39741-mimic
remote: Enumerating objects: 40, done.
remote: Counting objects: 100% (40/40), done.
remote: Total 50 (delta 40), reused 40 (delta 40), pack-reused 10
Unpacking objects: 100% (50/50), done.
From https://github.com/smithfarm/ceph
 * branch            wip-39741-mimic -> FETCH_HEAD
Auto-merging src/rgw/rgw_object_expirer_core.cc
Auto-merging src/rgw/rgw_admin.cc
Auto-merging doc/radosgw/dynamicresharding.rst
Auto-merging PendingReleaseNotes
CONFLICT (content): Merge conflict in PendingReleaseNotes
Automatic merge failed; fix conflicts and then commit the result.
Traceback (most recent call last):
  File "/home/yuriw/wip_master/src/script/build-integration-branch", line 62, in <module>
    assert not r
AssertionError

**Comment by yuriw** [comment]
[wip-yuri2-testing-2019-10-04-0015-mimic](https://trello.com/c/xfBj13mW/707-wip-yuri2-testing-2019-10-04-0015-mimic)

**Pull request merged**